### DISCUSSION TITLE
No API key found when login with OAuth (Twitter)

### DISCUSSION THREAD

**Discussion by josephlevarato** [supabase/supabase#15464] [Questions] [answered]
I'm trying to handle user login with Supabase and a Nuxt project.

I installed `@nuxtjs/supabase`, and followed the documentation:

- Created a Twitter API app
- Filled what was asked
- Provided Supabase with the Twitter secret and enabled Twitter as a provider,
- Added both `SUPABASE_URL` and `SUPABASE_KEY` to my .env file
- Added the following as a test to my code:

```
const supabase = useSupabaseAuthClient()
const login = async () => {
    const { data, error } = await supabase.auth.signInWithOAuth({
        provider: 'twitter',
        options: {
            redirectTo: 'https://example.com/welcome'
        }
    })

    if (error) {
        console.log({ error })
    }
}
```

When I trigger `login`, I get redirected to my supabase's url, without even seeing Twitter's login page, with the following error:

```
{"message":"No API key found in request","hint":"No `apikey` request header or url param was found."}
```

I have no clue of what I should do, I tried using the "classic" library (`@supabase/supabase-js`) but I got the same result. I checked 4/5 times that the api key was correct, still I get nothing.

What did I miss?

Thank you in advance

**Comment by ourkade-derek** [comment, marked as answer]
I had this same issue but managed to get it resolved. The current twitter auth instructions are a bit outdated and unclear.

When you are setting up your project, You MUST NOT click "Create App" or it will not have sufficient permissions:
![image](https://github.com/supabase/supabase/assets/149711759/e0199855-88d1-40eb-ae30-ad0e0517ff54)

Make sure you click "Add Project" to get started:
![image](https://github.com/supabase/supabase/assets/149711759/a41ebe45-ee1a-48f3-8e01-c85edf718437)

If you click "Create App", it will still create a project and add your app to it, but the permissions are different somehow.

The instructions say to "Turn Enable 3-legged OAuth ON." but this option doesn't exist anymore. I selected "Native App" under "Type of App" and it worked for me.

There doesn't seem to be a way to "reset" the application if it was created with the "Create App" button, so you'll need to delete it, then create a new project, then create an app underneath it.

**Reply by brianyang** [reply]
just to confirm, twitter auth is working for your supabase web app? you selected "native app", for a web app? 

i'm facing issues with twitter auth, but works with other providers. i suspect this bit of code is not working as expected

```
    const client = createRouteHandlerClient<Database>({ cookies });

    await client.auth.exchangeCodeForSession(code);
```

**Reply by eggplantlife** [reply]
Thank you @ourkade-derek, this saved us after many hours debugging!

**Reply by josephlevarato** [reply]
Hey, thanks! I forgot I even asked this question until I wondered again if I could login my users with Twitter / X. I still need to handle what happens after with [exchangeCodeForSession](https://supabase.com/docs/reference/javascript/auth-exchangecodeforsession) but the problem is fixed, thanks a lot!

**Comment by roborourke** [comment]
If it helps my issue was the same but the reason was that I had put the **oAuth client and secret** into the Supabase Twitter auth settings, rather than the **API key and secret** 🤦‍♂️
